Tools/ardupilotwaf/cmake.py

```diff
@@ -211,35 +207,103 @@ def _cmake_build_task_post_run(self):
-def process_cmake_configure(self):
-    if not hasattr(self, 'name'):
-        self.bld.fatal('cmake_configure: taskgen is missing name')
-    if not hasattr(self, 'cmake_src'):
-        self.bld.fatal('cmake_configure: taskgen is missing cmake_src')
+class CMakeConfig(object):
+    '''
+    CMake configuration. This object shouldn't be instantiated directly. Use
+    bld.cmake().
+    '''
+    def __init__(self, bld, name, srcnode, bldnode, cmake_vars):
+        self.bld = bld
+        self.name = name
+        self.srcnode = srcnode
+        self.bldnode = bldnode
+        self.vars = cmake_vars
+
+        self._config_task = None
+        self.last_build_task = None
+
+    def vars_keys(self):
+        keys = list(self.vars.keys())
+        if not isinstance(self.vars, OrderedDict):
+            keys.sort()
+        return keys
+
+    def config_sig(self):
+        m = Utils.md5()
+        def u(s):
+            m.update(s.encode('utf-8'))
+        u(self.srcnode.abspath())
+        u(self.bldnode.abspath())
+        keys = self.vars_keys()
+        for k in keys:
+            u(k)
+            u(self.vars[k])
+        return m.digest()
+
+    def config_task(self, taskgen):
+        sig = self.config_sig()
+        if self._config_task and self._config_task.cmake_config_sig == sig:
+            return self._config_task
+
+        # NOTE: we'll probably need to use the full class name in waf 1.9
+        self._config_task = taskgen.create_task('cmake_configure')
+        self._config_task.cwd = self.bldnode.abspath()
+        self._config_task.cmake = self
+        self._config_task.cmake_config_sig = sig
+
+        env = self._config_task.env
+        env.CMAKE_BLD_DIR = self.bldnode.abspath()
+        env.CMAKE_SRC_DIR = self.srcnode.abspath()
+
+        keys = self.vars_keys()
+        env.CMAKE_VARS = ["-D%s='%s'" % (k, self.vars[k]) for k in keys]
+
+        self._config_task.set_outputs(
+            self.bldnode.find_or_declare('CMakeCache.txt'),
+        )
 
-    if not isinstance(self.cmake_src, Node.Node):
-        self.cmake_src = self.bld.path.find_dir(self.cmake_src)
+        if self.last_build_task:
+            self._config_task.set_run_after(self.last_build_task)
 
-    self.get_cmake_bldnode()
-    self.cmake_bld.mkdir()
+        self.bldnode.mkdir()
 
-    self.last_build_task = None
+        return self._config_task
 
-    self.cmake_vars = getattr(self, 'cmake_vars', {})
+    def build(self, cmake_target, **kw):
+        return self.bld.cmake_build(self.name, cmake_target, **kw)
 
-    # NOTE: we'll probably need to use the full class name in waf 1.9
-    tsk = self.cmake_config_task = self.create_task('cmake_configure')
-    tsk.cwd = self.cmake_bld.abspath()
-    tsk.env.CMAKE_BLD_DIR = self.cmake_bld.abspath()
-    tsk.env.CMAKE_SRC_DIR = self.cmake_src.abspath()
+_cmake_instances = {}
+def get_cmake(name):
+    if name not in _cmake_instances:
+        bld.fatal('cmake: configuration named "%s" not found' % name)
```

Comment:
Oops... left over from a previous unstaged version of this function...
Thanks! I'll fix it

On Wed, May 25, 2016, 07:38 Francisco Ferreira <notifications@github.com>
wrote:

> In Tools/ardupilotwaf/cmake.py
> <https://github.com/ArduPilot/ardupilot/pull/4204#discussion_r64550701>:
>
> >
> > -    # NOTE: we'll probably need to use the full class name in waf 1.9
> > -    tsk = self.cmake_config_task = self.create_task('cmake_configure')
> > -    tsk.cwd = self.cmake_bld.abspath()
> > -    tsk.env.CMAKE_BLD_DIR = self.cmake_bld.abspath()
> > -    tsk.env.CMAKE_SRC_DIR = self.cmake_src.abspath()
> > +_cmake_instances = {}
> > +def get_cmake(name):
> > +    if name not in _cmake_instances:
> > +        bld.fatal('cmake: configuration named "%s" not found' % name)
>
> Does this work Gustavo? *bld* isn't received by the function
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/4204/files/2a71308a330d90a311961b0962b177aa5b6dd718#r64550701>
>